blocks/loader.ts

```diff
@@ -57,24 +68,145 @@ export const wrapCaughtErrors = async <
     });
   }
 };
+
+export const ENABLE_LOADER_CACHE =
+  Deno.env.get("ENABLE_LOADER_CACHE") !== undefined;
+
+const stats = {
+  cache: meter.createCounter("loader_cache", {
+    unit: "1",
+    valueType: ValueType.INT,
+  }),
+  latency: meter.createHistogram("resolver_latency", {
+    description: "resolver latency",
+    unit: "ms",
+    valueType: ValueType.DOUBLE,
+  }),
+};
+
+let maybeCache: Promise<unknown> | Cache | undefined = caches.open("loader")
+  .then((c) => maybeCache = c)
+  .catch(() => maybeCache = undefined);
+
+const MAX_AGE_S = 60; // 60 seconds
+
+const isCache = (c: any): c is Cache => typeof c?.put === "function";
+
+const inFuture = (maybeDate: string) => {
+  try {
+    return new Date(maybeDate) > new Date();
+  } catch {
+    return false;
+  }
+};
+
+const noop = () => "";
+
+const wrapLoader = ({
+  default: handler,
+  cache: mode = "no-store",
+  cacheKey = noop,
+  singleFlightKey,
+  ...rest
+}: LoaderModule) => {
+  const flights = singleFlight();
+
+  if (typeof singleFlightKey === "function") {
+    console.warn(
+      "singleFlightKey is deprecated and does not work anymore. Please use cacheKey instead",
+    );
+  }
+
+  return {
+    ...rest,
+    default: async (
+      props: Parameters<typeof handler>[0],
+      req: Request,
+      ctx: FnContext<unknown, any>,
+    ): Promise<ReturnType<typeof handler>> => {
+      const loader = ctx.resolverId || "unknown";
+      const start = performance.now();
+      const skipCache = mode === "no-store" || !ENABLE_LOADER_CACHE ||
+        !isCache(maybeCache);
+
+      let status: "bypass" | "miss" | "stale" | "hit" | undefined;
+
+      try {
+        if (skipCache) {
+          status = "bypass";
+          stats.cache.add(1, { status, loader });
+
+          return await handler(props, req, ctx);
+        }
+
+        // Somehow typescript does not understand maybeCache is Cache
+        const cache = maybeCache as Cache;
+
+        // TODO: Resolve props cache key statically
+        const key = `${hash(props)}-${cacheKey(req, ctx)}`;
+        const request = new Request(
+          `https://localhost?propsKey=${hash(props)}&requestKey=${key}`,
```

Comment:
add a comment explaning why localhost (I know that its just an arbitrary cache key and the host doesn't matter but it might worth to let it explicit)